File: C++/Learing-C++/20. C++ OOPs/3. Encapsulation/encapsulation.cpp
#include <iostream>
#include <string>

using namespace std;

class Student {
private:
    // Private member variables
    string name;
    int age;

public:
    // Setter for name
    void setName(string newName) {
        name = newName;
    }
    // Getter for name
    string getName() const {
        return name;
    }

    // Setter for age with validation
    void setAge(int newAge) {
        // Validate and set age
        if (newAge >= 0) {
            age = newAge;
        } else {
            cout << "Invalid age!\n";
        }
    }

    // Getter for age
    int getAge() const {
        return age;
    }

    // Display information
    void displayInfo() const {
        cout << "Name: " << name << "\n";
        cout << "Age: " << age << "\n";
    }
};

int main() {
    // Create an instance of the Student class
    Student student1;

    student1.setName("John");
    student1.setAge(20);

    cout << "Student Information:\n";
    student1.displayInfo();

    return 0;
}

File: Dart/Smart Dart/16. Dart OOPs/04. Encapsulation/03. Encapsulation Using Getter&Setter Method/encapsulation.dart
import 'person.dart';

void main() {
  // Creating a new Person object & using cascade notation (..) to set the 'name' property
  Person person = Person()..name = "Sadman";

  // Accessing public field and method
  print("Name: ${person.name}");

  // Using setter to update private field
  person.setAge = 20;

  // Accessing private field and using getter
  print("Age: ${person.getAge}");

  // Calling the introduceYourself method to display information about the person
  person.introduceYourself();
}


/*

  In Dart, encapsulation is achieved through the use of access modifiers to control
  the visibility of class members. This example demonstrates encapsulation without
  using explicit getter and setter methods.

  Author: [Your Name]
  Date: [Current Date]

  [Class: Person]
  - The 'Person' class encapsulates a person's information with private instance variables:
    'String? _name' and 'int? _age'.
  - The constructor initializes the private variables.
  - Public methods are provided to access and update the private information.

  [Main Function]
  - In the 'main' function, an instance of the 'Person' class is created.
  - The public methods are used to access and update the person's information.

  This example illustrates how encapsulation helps protect the internal state of
  an object by restricting direct access to its private members, promoting data integrity.
  
*/


